### ISSUE TITLE
Daily Liveweight Gain graph on reports

### ISSUE THREAD

**Issue by Skipper-is** [farmOS/farmOS#307]
The weight funciton is really great, but there is a slight, bug? Well, unhelpfulness, in the report generated.
![image](https://user-images.githubusercontent.com/28678069/83972468-32fc0180-a8d8-11ea-863d-735497035a0b.png)

The bit I'm looking at specifically is the average daily gain.
If all animals are weighed on the same day - the current graph would probably work, but that is not always the case!
Currently the daily gain is taking the previous weight, so for the last 2 records in the graph - One goat on 14th May at 5.3kg, and then the next record of a number of other goats on 3rd June at 21kg. Average daily gain is reporting at 0.7866kg/day - which is far too high! That suggests that the individual on the 14th (5.3kg) now weighs 21kg, which would be record breaking!

What would be much more helpful would be if it tracks individual daily liveweight gains, and then projects the average of that.

So for example - My 5.3kg goat on the 14th May weighs 9kg on the 3rd June. 3.4kg dif/20 days = 0.17kg/day.  On the 20th, it's daily liveweight gain is 0.17kg.
My 21kg one was born on the 13/04/2020, a month earlier (at 5kg), so it's daily liveweight gain is 0.313kg/day. 
For the graph, we just need to plot the average of those two. 

Unfortunately, php isn't my main language, so I can only really explain this concept in Python - 

If each animal has an array of weights, and a matching array of dates:
The graph needs to be something along the lines of:
```
weight_difference = animal_a[-1]-animal_a[-2]
date_difference = animal_a_date[-1]-animal_a_date[-2]

daily_liveweight_gain = weight_difference/date_difference
```

Repeat for each animal in the group, and then average it.
Something more like this:
![image](https://user-images.githubusercontent.com/28678069/83973280-2fb74480-a8dd-11ea-89de-3d0a34146c5f.png)
From this data:
|Name  | 20/03 |21/03 |28/03|14/05|03/06|
|--|--|--|--|--|--|
|Oliver||||5.6kg|9kg (+0.17kg/d)|
|Ozzy||4.8kg|||28kg (+0.31kg/d)
|Octavita||4.7kg|||28kg (+0.32kg/d)
|Osian| 5.2kg||7.2kg (+0.24kg/d)||30kg (0.34kg/d)
|||||Average:|+0.285kg/d

**Comment by paul121** [comment]
Thanks for documenting this, @Skipper-is ! My apologies for not getting back sooner like I said I would in chat!

Like you said, the crux of the problem seems to be that the report assumes animals are weighed on the same day. Initially that worked fine for the first use case of the report (random or complete sample of animals weighed on the same day).

But now they too are outgrowing it's current capabilities. I'm working on adding some additional filter features (#323) that help narrow the report down to only specific assets (filter by asset flags & birth date range). This helps to narrow the report down to a few assets or even a single asset. After filtering, it's more likely to fit the current criteria that "animals are weighed on the same day" - but filters alone won't solve this issue! This is only a first step towards more accurate reporting of individual assets.

One thing I said in chat...
> I'm curious if we might need some different "modes". I think there still is a use case where a random sampling of animals are weighed. So individual assets might not be tracked

After thinking about this more, I'm curious if we don't need different modes. It seems like calculating average weights + gains on an individual level (as you described) and then averaging over a period of time with other assets in the group would provide the same result for the "random" sampling method. Perhaps the ability to select groups of assets OR type in multiple asset names would cover the "modes" I was thinking about.

I think there is another piece to figure out though: what is the time period we are averaging?

@Skipper-is your graph looks like it's just averaging everyones weights from 20/03 to 03/06. Since they mostly only have 2 recorded weights, this seems OK. But... what happens when each individual has 4-5 weights recorded? Averaging from the beginning to the end of the season will give a flat line, but doesn't capture that the animals may be gaining weight slower or faster throughout the season. (The current implementation captures this in the graph since all animals are weighed on the same date, so changes in slope from date to date can be seen)

It seems like we might need to average all weights + gains at a set interval over the report. Eg: for each interval (each day, every 3 days, weekly, etc) average all weights + gains for animals at that point in time. Calculating this multiple times throughout the report would give better insight to how the group as a whole is changing over time. Depending on the interval, this might result in lots of calculations... but we can probably optimize that.

An additional option that comes to mind: only include animals' first X or last X number of recorded weights. Alternatively, only display animals' first X or last X number of days in the report.

<hr>

TLDR; I think we could switch to calculating average weights + gains on an individual level as @Skipper-is described, but need to figure out a way to see changes over time. Adding some kind of a calculation "interval" comes to mind, which could even be configurable in the report input form.

Also we should consider adding an option to select an asset group OR type in individual assets to average / compare.

**Comment by paul121** [comment]
At the end of that comment I mentioned the ability to "compare" individual assets... that would be a bit more complicated to add (graphing multiple individuals/groups on the same graph) BUT it gets me thinking... 

Perhaps the weight report could be a "display only" Vue.js app that manages the graphign of averages + gains. This would allow for some more "interactive" features in the report. Specifically if all recorded weights for the selected assets were provided to the "app" on page load. Some ideas:

- The date range & "intervals" described above could be calculated on the fly without an additional page load.
- Graph individual assets along side the groups average (see individuals that are growing faster/slower)
- Compare individual assets to one another
- Easily exclude individual assets from the average calculation (if there are outliers, animals that aren't in a group for a period of time, etc)

**Comment by Skipper-is** [comment]
I've not found any similar graphs from other farm management software, so I'm a bit in the dark really on other ways of presenting the data.
At the minute my fork/pull which gives dlwg for each individual does the job I need. But I'm not sure how to present rhe data for all animals in a meaningful way.
Some of the ways I've seen it presented are with DLWG Vs age, with a different line being each individual, but that would get cluttered fast, and requires quite a lot of data, and doesn't work well on random sampling.

**Comment by jgaehring** [comment]
Hi all,

> the crux of the problem seems to be that the report assumes animals are weighed on the same day

Is it that, or is it that the number of animals being weighed is not constant? I feel like the time interval is incidental to the data being graphed. Seems like what you really need is the average of the per capita weight, not the average of the total weight. That would introduce its own set of fluctuations, to be sure, unless you somehow factored in the specific animals being weight, but I feel like it would still give a better representation of the salient features of the data set.

Also, I wonder if a line chart is the best form of visualization here. Seems like a scatter plot and linear regression might be better suited. So it would look something like this:

[![linear regression](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Linear_regression.svg/438px-Linear_regression.svg.png)]

Perhaps the time intervals could come into play in segmenting the regression line?

Sorry if this is totally off the mark, I'm just getting to speed on all this.

"Came for the JavaScript, stayed for the data viz." ~ me

**Comment by Skipper-is** [comment]
What would our X axis be? Date? Then it wouldn't show +/- gains very well over different individuals of different age/stage - would just be a cloud... I do think age might be the best to compare against

**Comment by Skipper-is** [comment]
Or weigh gain Vs total weight I suppose..

**Comment by jgaehring** [comment]
I guess it depends on what features you really want to stand out. I was thinking in terms of overall health of the herd/flock, like if you moved them to new pasture between a certain set of dates and they didn't put on as much weight, you could potentially see that in the viz.

But if you are more interested in being able to flag a certain individual from the herd, as having an unhealthy growth trend, then maybe something like sparklines?

![image](https://user-images.githubusercontent.com/5556545/87803905-2943b300-c821-11ea-9a04-c68f3e9263a2.png)

I'm just spitballing here, but it does seem to me that the original line chart, as formulated, might not be the best way of visualizing all the relevant dimensions to the data and their potential derivatives.

**Comment by mstenta** [comment]
I'm closing old/stale GitHub issues. We have a newer issue opened in the farmOS Animal Weight module regarding daily liveweight gain. I linked to this thread from that issue so we don't lose all this great discussion.

https://www.drupal.org/project/farm_animal_weight/issues/3240954
